### ISSUE TITLE
Retention policies improvement

### ISSUE THREAD

**Issue by jay1337** [gilbertchen/duplicacy#277]
Current implementation of retention policies keeps a revision if (and only if):

- it is the oldest for the current active policy
- or, it's not too close to the revision just older kept

I think that a revision must not be removed if removing it makes the time between 2 successive revisions greater than what is required by the retention policy.
With "-keep 7:1" and these revisions:

1. 01/01/2000 12:00:00
2. 08/01/2000 11:00:00
3. 15/01/2000 12:00:00

The current implementation will delete revision 2 so that there will be 14 days between the remaining revisions 1 and 3. Not good...

I suggest a new implementation that would keep a revision if (and only if):

- it is the oldest for the current active policy
- or, it's not too close to the revision just older kept
- or, time between previous revision (the "just older kept" ie "the most recent kept older than the current" cf current variable "lastSnapshotTime") and the next (the "just more recent" ie "the oldest existing more recent than the current") is lower than the interval of the retention policy

It would make the current 10min tolerance useless.

What do you think about that ?

**Comment by gilbertchen** [comment]
I agree this should be improved.  I went for the current implementation only because it was the simplest to implement.
